### ISSUE TITLE
Google for WooCommerce Action Scheduler queue becomes stuck/slows dramatically during product sync

### ISSUE THREAD

**Issue by byontea** [woocommerce/google-listings-and-ads#3729]
### Describe the bug:

After a bulk product update, Google for WooCommerce creates a large number of `gla/jobs/update_products/process_item` actions, each containing a single product ID.

On one of our WooCommerce stores, approximately **205–206 actions became past due at the same time** and Action Scheduler started reporting:

`The maximum number of concurrent batches is already running (1 batch). No additional batches will be processed until the current batch completes.`

At the same time, our hosting monitoring showed repeated slow requests to:

`/wp-admin/admin-ajax.php?action=as_async_request_queue_runner`

with average execution times between approximately **43 and 59 seconds**, and maximum execution times up to **101 seconds**.

The affected actions remained in `Pending / Past Due` status for a long period.

Some of them were later automatically reset by Action Scheduler, with log entries such as:

`action reset`

After some time, the queue eventually started processing again.

We have another very similar WooCommerce store, with approximately 10,000 products, also using Google for WooCommerce. That store also creates individual `gla/jobs/update_products/process_item` actions containing one product ID, but those actions normally complete within **0–2 seconds**.

Therefore, the issue does not appear to be caused simply by the number of products or by the fact that one action is created per product. The problematic store appears to get into a state where GLA actions become extremely slow, causing the single Action Scheduler queue to remain occupied and hundreds of actions to become past due.

### Steps to reproduce:

1. Perform a bulk update of a significant number of WooCommerce products, such as stock or price updates.

2. Open WooCommerce → Status → Scheduled Actions and monitor `gla/jobs/update_products/process_item`.

3. Observe that a large number of individual actions are created at approximately the same time, each containing one product ID, for example:

   `[[15099]]`
   `[[15091]]`
   `[[5570]]`

4. On the affected store, observe that many of these actions become `Past Due`.

5. Action Scheduler may then display:

   `The maximum number of concurrent batches is already running (1 batch). No additional batches will be processed until the current batch completes.`

6. Server monitoring shows long-running requests to:

   `/wp-admin/admin-ajax.php?action=as_async_request_queue_runner`

   with execution times of tens of seconds and, in our case, up to approximately 101 seconds.

7. Some claimed actions may later be reset back to pending state by Action Scheduler.

### Expected behavior:

`gla/jobs/update_products/process_item` actions should be processed quickly enough that bulk product updates do not create a large Action Scheduler backlog.

Even if hundreds of individual product update actions are generated, the queue should continue draining normally without keeping the single Action Scheduler batch occupied for 40–100+ seconds.

Ideally, single-product updates generated within a short time window could also be coalesced so that the existing Merchant API batching can process multiple products together.

### Actual behavior:

On the affected store, approximately 205–206 GLA product update actions became past due.

The Action Scheduler queue became effectively blocked by the single concurrent batch, while new actions continued becoming past due.

Observed Action Scheduler counts during the issue included approximately:

* Completed: 34,812
* Failed: 257
* Pending: 227
* Past Due: 206
* Cancelled: 15

Most affected GLA actions were created at approximately:

`2026-08-24 14:00:09 UTC`

The hosting monitor reported repeated slow Action Scheduler async requests with examples such as:

* 6 requests — 59 s average — 101 s maximum
* 20 requests — 50 s average — 72 s maximum
* 6 requests — 48 s average — 62 s maximum
* 7 requests — 43 s average — 58 s maximum

Some actions were later automatically reset and returned to pending status.

Eventually the queue began processing again, but only after a significant backlog and period of degraded performance.

It would be useful to understand why `gla/jobs/update_products/process_item` can occasionally take tens of seconds on one store while identical jobs on another similar store complete in 0–2 seconds, and whether cross-request coalescing could reduce the impact of these slow periods.

**Comment by builtbytay** [comment]
@byontea  Thanks for the thorough write-up — the comparison against your other ~10k store is especially useful, since it rules out raw product count.

Your read looks right. Action Scheduler ends a batch once the projected time would exceed its 30-second limit, so a queue of 0–2s actions would finish its runner request around 25–30s — not 50s, not 101s. Requests running 43–101s are hard to explain unless individual actions really are taking tens of seconds, which is what you suspected. The open question is which ones, and what they're waiting on. Would you be able to provide a few things to help us narrow it down? 

**1. Per-action timings.** In WooCommerce → Status → Scheduled Actions, filter to `gla/jobs/update_products/process_item` and screenshot the Log column for a few affected actions. It timestamps "action started" and "action complete", so the gap is the real duration — and it'll show whether the slowness is spread evenly or concentrated in a few. Worth grabbing soon: completed actions are purged after about a month, so the 2026-08-24 batch ages out in late September.

**2. Whether the slow ones are variable products.** A variable product schedules one action carrying *all* its variations — so 40 variations means one action doing 40 syncs, which can legitimately run long. Slow actions with long argument arrays, rather than the single `[[15099]]` shape you quoted, would point straight at this.

**3. Your system report** (WooCommerce → Status). Version matters — we have a separate report of outbound requests timing out on 3.9.0, which would produce this exact symptom. Also useful: what drives the bulk update — bulk edit, CSV import, REST, or a plugin.

On the coalescing idea at the end of your report: that's tracked at #3624, from another merchant hitting the same one-action-per-product pattern via an ERP. Worth following, though based on the above it's probably not what's impacting your site.

**Comment by byontea** [comment]
Thanks — I checked the Scheduled Actions in more detail and found something quite significant.

The very long runner requests appear to be caused by individual GLA actions, rather than simply a large queue.

The slow actions I found are:

`gla/jobs/update_merchant_product_statuses/process_item`

Examples from today (UTC):

* Started 06:20:03 → completed 06:25:28: **~325s**
* Started 06:30:04 → completed 06:35:26: **~322s**
* Started 06:40:04 → completed 06:45:21: **~317s**
* Started 06:45:25 → completed 06:51:01: **~336s**

For the first three, Action Scheduler marked them as failed after 300 seconds with:

`action was in-progress for at least 300 seconds without completing and has been marked as failed`

However, they subsequently completed successfully a few seconds later.

For example:

`06:40:04 action started via WP Cron`
`06:45:04 marked failed after 300 seconds`
`06:45:21 action completed via WP Cron`

The arguments for these actions are not product/variation arrays. They only contain a pagination token, for example:

`0 => array ( 'next_page_token' => 'CITPAQ==', )`

Other consecutive actions contain different `next_page_token` values, so this appears to be the Merchant product-status pagination process rather than a single large variable product.

This also correlates very closely with the slow HTTP request I see on the server:

`/wp-admin/admin-ajax.php?action=as_async_request_queue_runner`

Request count: 2
Max time: **342s**
Average time: **173s**

So at least in this case there is definitely an individual GLA action running for more than five minutes.

Regarding the other points:

* Google for WooCommerce is on the **latest currently available version**.
* MySQL slow-query logging shows **no slow queries during the affected period**.
* The PHP/server error log also does not show a corresponding PHP fatal error or execution-timeout error during those actions.

I also checked the Google for WooCommerce debug log. Earlier in the day I did see some outbound requests returning:

`401 Unauthorized`
`Invalid token`
`Please reconnect your Jetpack account`

However, later GLA Merchant API batches were also completing successfully (for example, batches of 100 products with 100 succeeded / 0 failed), so I am not yet sure whether the authentication issue is related to these 300+ second `update_merchant_product_statuses` actions.

The interesting part now seems to be determining what `gla/jobs/update_merchant_product_statuses/process_item` is waiting on for ~320–340 seconds — potentially an outbound Merchant API request.

I can also provide the full WooCommerce System Status Report and screenshots of the Scheduled Action logs if useful.

**Comment by builtbytay** [comment]
@byontea Thank you! This resolves the open question, and it's the confirmation we were missing.

Confirmed code bug, and a duplicate of an existing report already being defined. The cause: the HTTP client used for Merchant API calls is created with no timeout, so a slow-but-successful response from Google has no ceiling — the action waits until Action Scheduler kills it at 300s. Your log is the clearest evidence of this we've had: started 06:40:04, marked failed 06:45:04, completed 06:45:21. It was never stuck, just slow — which is why your PHP and MySQL logs are clean.

On your two open points: the `Invalid token` errors are a separate signal, since auth failures throw immediately rather than stalling; and your healthy ~10k store fits this, because the trigger is Google's response latency, not your catalog.

I'm going to close this as a duplicate and carry your report over to the active issue, which already covers bounding the wait and keeping one stalled action from taking others down.
